staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go

```diff
@@ -355,14 +355,66 @@ type ListOptions struct {
 	// +optional
 	AllowWatchBookmarks bool `json:"allowWatchBookmarks,omitempty" protobuf:"varint,9,opt,name=allowWatchBookmarks"`
 
-	// When specified with a watch call, shows changes that occur after that particular version of a resource.
-	// Defaults to changes from the beginning of history.
-	// When specified for list:
-	// - if unset, then the result is returned from remote storage based on quorum-read flag;
-	// - if it's 0, then we simply return what we currently have in cache, no guarantee;
-	// - if set to non zero, then the result is at least as fresh as given rv.
+	// ResourceVersion sets a constraint on what resource versions a request may be served from.
+	// If unset, the request is served from latest resource version to ensure strong consistency
+	// (i.e. served from etcd via a quorum read). Setting a ResourceVersion is preferable in cases
+	// where a ResourceVersion is known and can be used as a constraint since better performance
+	// and scalability can be achieved for a cluster by avoiding quorum reads.
+	// When ResourceVersion is set for list, it is highly recommended ResourceVersionMatch is also set.
+	//
+	// watch semantics:
+	// - if unset, start a watch at the most recent resource version, which must be consistent
```

Comment:
That isn't what we do now. In fact for watch, "" and "0" are not differentiated and treated exactly the same:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go#L436

Which basically means we have not fully consistent semantics between watchcache and pure etcd3 client implementation.

That said - I'm not sure how bad that is - actually I would love to deprecate and remove the ability of passing RV="0" and RV="" to watch (and actually force users to start from an explicitly given RV, which all our components are doing anyway unless I'm missing something) - maybe it's actually time to do that?

@liggitt @lavalamp @smarterclayton ^

Reply:
Looks like we documented how the watch cache handles this in https://kubernetes.io/docs/reference/using-api/api-concepts/#the-resourceversion-parameter wrong too.

I'll update this comment to document that unset has the same watch guarantee as "0" so that this is not misleading. I'll send out a documentation update once we're in agreement here.

Reply:
A slight inconsistency between behavior of the watch cache and direct-to-etcd in this scenario is likely not even noticed by the clients starting watches this way... most are just starting watch streams without expectations about a particular starting point.

> I would love to deprecate and remove the ability of passing RV="0" and RV="" to watch (and actually force users to start from an explicitly given RV, which all our components are doing anyway unless I'm missing something) - maybe it's actually time to do that?

even though we don't have any in-tree components using this, I'm positive that removing support for it would break lots of clients.

> I'll update this comment to document that unset has the same watch guarantee as "0" so that this is not misleading. I'll send out a documentation update once we're in agreement here.

we should definitely update the docs to reflect reality, independent of the discussion here

Reply:
Thanks @liggitt. I'll open the documentation fix.